src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs

```diff
@@ -196,55 +199,64 @@ public bool TryModifyEntries(ImmutableArray<T> outputs, IEqualityComparer<T> com
 
-                var previousEntries = _previous._states[_states.Count];
-                var modifiedEntries = ArrayBuilder<(T item, EntryState state)>.GetInstance();
+                var previousEntry = _previous._states[_states.Count];
+                var modifiedEntries = ArrayBuilder<T>.GetInstance();
+                var modifiedStates = ArrayBuilder<EntryState>.GetInstance();
 
-                var previousEnumerator = previousEntries.GetEnumerator();
+                var previousEnumerator = previousEntry.Items.GetEnumerator();
                 var outputEnumerator = outputs.GetEnumerator();
 
-                bool previousHasItems = previousEnumerator.MoveNext();
+                bool previousHasItems = previousEntry.IsSingle || previousEnumerator.MoveNext();
                 bool outputHasItems = outputEnumerator.MoveNext();
 
                 // cached or modified items
                 while (previousHasItems && outputHasItems)
                 {
-                    var previous = previousEnumerator.Current;
+                    var previous = previousEntry.IsSingle ? previousEntry.Item : previousEnumerator.Current;
                     var replacement = outputEnumerator.Current;
 
-                    var entryState = comparer.Equals(previous.item, replacement) ? EntryState.Cached : EntryState.Modified;
-                    modifiedEntries.Add((replacement, entryState));
+                    var entryState = comparer.Equals(previous, replacement) ? EntryState.Cached : EntryState.Modified;
+                    modifiedEntries.Add(replacement);
+                    modifiedStates.Add(entryState);
 
-                    previousHasItems = previousEnumerator.MoveNext();
+                    previousHasItems = !previousEntry.IsSingle && previousEnumerator.MoveNext();
                     outputHasItems = outputEnumerator.MoveNext();
                 }
 
                 // removed
                 while (previousHasItems)
                 {
-                    modifiedEntries.Add((previousEnumerator.Current.item, EntryState.Removed));
+                    modifiedEntries.Add(previousEntry.IsSingle ? previousEntry.Item : previousEnumerator.Current);
+                    modifiedStates.Add(EntryState.Removed);
                     previousHasItems = previousEnumerator.MoveNext();
                 }
 
                 // added
                 while (outputHasItems)
                 {
-                    modifiedEntries.Add((outputEnumerator.Current, EntryState.Added));
+                    modifiedEntries.Add(outputEnumerator.Current);
+                    modifiedStates.Add(EntryState.Modified);
                     outputHasItems = outputEnumerator.MoveNext();
                 }
 
-                _states.Add(modifiedEntries.ToImmutableAndFree());
+                _states.Add(new TableEntry(modifiedEntries.ToImmutableAndFree(), modifiedStates.ToImmutableAndFree()));
```

Comment:
If you agree with my idea of having a singleton `ImmutableArray<EntryState>` for each of the core states then I think we could actually make this even more efficient. Consider that this code path is going to build up `ImmutableArray<EntryState>` values where the length is greater than one but all the `EntryState` are the same value. It should be easy to write a simple builder here that can catch that case and re-use the singleton arrays. 

That would simplify some of the checks in `TableEntry` and `NodeStateTable` because we could assume that in the case we had a `ImmutableArray<EntryState>` where the length is greater than one that it was heterogeneous.

Reply:
Ah, I think you're saying that we assume its homogenous, until it's not? At which point we'd create an array to hold the differences. 

I wonder if we could just create a lookup in that case? Essentially a dictionary of index -> state, where no entry indicates the item has a state of `_state`. Honestly not sure of the perf of this vs. the arrays of index one though tbh.

Reply:
Yes. Essentially I imagine you can store a tuple of `(EntryState State, int Count)` and at the moment the next state doesn't match the `State` item then you flip to an `ArrayBuilder<EntryState>` and populate with `Count` items then add the next item. But if we get to the end and haven't created a builder we feed into the singleton case